### PR TITLE
Only warn about unsupported bwd sync if enabled

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

If your fabric script supports multiple strategies and it integrates the `no_backward_sync` context manager, fabric shouldn't show this warning if the strategy doesn't support it but it's not enabled

----
📚 Documentation preview 📚: https://pytorch-lightning--19709.org.readthedocs.build/en/19709/

### CODE REVIEW

File: src/lightning/fabric/fabric.py [hunk 1]
```diff
@@ -674,7 +674,7 @@ def no_backward_sync(self, module: _FabricModule, enabled: bool = True) -> Conte
             )
         if isinstance(self._strategy, (SingleDeviceStrategy, XLAStrategy)):
             return nullcontext()
-        if self._strategy._backward_sync_control is None:
+        if enabled and self._strategy._backward_sync_control is None:
```
Comment (awaelchli) [inline src/lightning/fabric/fabric.py:677]: I vote for keeping it as is, because the enabled bool here is typically not a constant. In the case of gradient accumulation of N steps, it is first False during N-1 iterations and then True for 1 iteration. With this change, the warning would be delayed to the N-th iteration, isn't that odd?

Comment (carmocca) [reply src/lightning/fabric/fabric.py:677]: Let me tell you why I did this and maybe you agree with me. I am using the litgpt training script (which uses `no_backward_sync`) to run two strategies. One that supports no sync and one that doesnt.

For an apple-to-apples comparison, I tweaked the batch size in the config so that neither of them enable gradient accumulation (`enabled=False` always).

However, this warning still appeared and that made me think I had made a mistake when I didn't.

I prefer my suggestion because in your case, even though it's delayed by N-1 steps, it still does the correct thing (warn about it). In my case, the warning is completely unnecessary

**Review by awaelchli — APPROVED** [review]
